### PR TITLE
discovery(engine): refactor agent to use list api

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tiflow/issues/6867

### What is changed and how it works?

Refactor `discovery.Agent` to use `ListMasters` and `ListExecutors` API.

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: engine/executor/server.go [hunk 1]
```diff
@@ -751,13 +721,18 @@ func (s *Server) bgUpdateServerMasterClients(ctx context.Context) error {
 		select {
 		case <-ctx.Done():
 			return perrors.Trace(ctx.Err())
-		case info, ok := <-s.cliUpdateCh:
-			if !ok {
-				// s.cliUpdateCh is closed.
-				return nil
+		case <-time.After(defaultDiscoveryAutoSyncInterval):
```
Comment (liuzix) [inline engine/executor/server.go:724]: Please consider using a `Ticker` unless it's absolutely necessary to use `time.After`. The reason is that `time.After` can leak timers.

File: engine/internal/pkg/discovery/agent.go [hunk 2]
```diff
@@ -38,182 +35,166 @@ const (
...
-				zap.Duration("duration", duration))
+func (a *agentImpl) autoSync(ctx context.Context) error {
+	snap := make(Snapshot)
+	for {
+		newSnap, err := a.getSnapshot(ctx)
+		if err != nil {
+			if errors.Cause(err) == context.Canceled || errors.Cause(err) == context.DeadlineExceeded {
+				return errors.Trace(err)
+			}
+			log.Warn("failed to get snapshot", zap.Error(err))
```
Comment (liuzix) [inline engine/internal/pkg/discovery/agent.go:144]: When can this happen?

File: engine/internal/pkg/discovery/agent.go [hunk 3]
```diff
@@ -38,182 +35,166 @@ const (
...
+	defer eventNotifier.Close()
 
-	discoveryCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	syncErr := make(chan error, 1)
+	syncDone := make(chan struct{})
+	go func() {
+		defer close(syncDone)
+		syncErr <- a.autoSync(ctx)
+	}()
```
Comment (liuzix) [inline engine/internal/pkg/discovery/agent.go:108]: Do you consider using errgroup?

File: engine/executor/server.go (hunk 1 above)
Comment (gozssky) [reply engine/executor/server.go:724]: `Ticker` will ignore the request latency. Is it possible to leak the timer every time it is created, or just when clicking "<-ctx.Done()".

File: engine/internal/pkg/discovery/agent.go (hunk 2 above)
Comment (gozssky) [reply engine/internal/pkg/discovery/agent.go:144]: When encountering network error or servermaster failover.

File: engine/internal/pkg/discovery/agent.go (hunk 3 above)
Comment (gozssky) [reply engine/internal/pkg/discovery/agent.go]: Changed to use errgroup.

File: engine/internal/pkg/discovery/agent.go [hunk 4]
```diff
@@ -38,182 +36,160 @@ const (
...
 
+// NodeType describes the type of the node.
+type NodeType string
+
+const (
+	// NodeTypeMaster indicates that the node is a master.
+	NodeTypeMaster = NodeType("master")
+	// NodeTypeExecutor indicates that the node is an executor.
+	NodeTypeExecutor = NodeType("executor")
+)
```
Comment (amyangfei) [inline engine/internal/pkg/discovery/agent.go:47]: Why don't reuse `NodeType` defined in engine/model/cluster.go

Comment (gozssky) [reply engine/internal/pkg/discovery/agent.go:47]: `NodeInfo` and `NodeType` defined in `engine/model/cluster.go` will be removed after we store executor info in MySQL.

**Review by amyangfei — APPROVED** [review]

**Review by liuzix — APPROVED** [review]

**Pull request merged**